rdmalib/include/rdmalib/buffer.hpp

```diff
@@ -2,192 +2,396 @@
-    Buffer(T * ptr, uint32_t size):
-      impl::Buffer(ptr, size, sizeof(T))
-    {}
+    Buffer(T *ptr, uint32_t size) : ImplBuffer(ptr, size, sizeof(T))
+    {
+    
+    }
 
     // Provide a buffer instance for existing memory pool
     // Does NOT free the associated resource
-    Buffer(void * ptr, uint32_t size):
-      impl::Buffer(ptr, size, sizeof(T))
-    {}
+    Buffer(void *ptr, uint32_t size) : ImplBuffer(ptr, size, sizeof(T))
+    {
+    }
 
-    Buffer(size_t size, size_t header = 0):
-      impl::Buffer(size, sizeof(T), header)
-    {}
+    Buffer(size_t size, size_t header = 0) : ImplBuffer(size, sizeof(T), header)
+    {
+    }
 
-    Buffer<T> & operator=(Buffer<T> && obj)
+    Buffer<T, Library> &operator=(Buffer<T, Library> &&obj)
     {
-      impl::Buffer::operator=(std::move(obj));
+      ImplBuffer::operator=(std::move(obj));
       return *this;
     }
 
-    Buffer(const Buffer<T> & obj) = delete;
-    Buffer(Buffer<T> && obj) = default;
+    Buffer(const Buffer<T, Library> &obj) = delete;
+    Buffer(Buffer<T, Library> &&obj) = default;
 
-    T* data() const
+    T *data() const
     {
       // void pointer arithmetic is not allowed
-      return reinterpret_cast<T*>(static_cast<char*>(this->_ptr) + this->_header);
+      return reinterpret_cast<T *>(static_cast<char *>(this->_ptr) + this->_header);
     }
   };
 
-  struct ScatterGatherElement {
-    // smallvector in practice
-    #ifdef USE_LIBFABRIC
-    mutable std::vector<iovec> _sges;
-    mutable std::vector<void *> _lkeys;
-    #else
-    mutable std::vector<ibv_sge> _sges;
-    #endif
+  template <typename Derived, typename Library>
+  struct ScatterGatherElement
+  {
+    using sge_t = typename library_traits<Library>::sge_t;
+    using lkey_t = typename library_traits<Library>::lkey_t;
+
+    mutable std::vector<sge_t> _sges;
 
-    ScatterGatherElement();
+    ScatterGatherElement()
+    {
+    }
 
-    #ifdef USE_LIBFABRIC
-    ScatterGatherElement(uint64_t addr, uint32_t bytes, void *lkey);
-    #else
-    ScatterGatherElement(uint64_t addr, uint32_t bytes, uint32_t lkey);
-    #endif
+    ScatterGatherElement(uint64_t addr, uint32_t bytes, lkey_t lkey)
+    {
+      static_cast<Derived *>(this)->Derived(addr, bytes, lkey);
+    }
 
-    template<typename T>
-    ScatterGatherElement(const Buffer<T> & buf)
+    template <typename T>
+    ScatterGatherElement(const Buffer<T, Library> &buf)
     {
       add(buf);
     }
 
-    template<typename T>
-    void add(const Buffer<T> & buf)
+    template <typename T>
+    void add(const Buffer<T, Library> &buf)
+    {
+      static_cast<Derived *>(this)->add(buf);
+    }
+
+    template <typename T>
+    void add(const Buffer<T, Library> &buf, uint32_t size, size_t offset = 0)
+    {
+      static_cast<Derived *>(this)->add(buf, size, offset);
+    }
+
+    sge_t *array() const
+    {
+      return static_cast<Derived *>(this)->array();
+    }
+
+    size_t size() const;
```

Comment:
Shouldn't this function be defined here? A function member of a template class should not go into a .cpp file IMO.

Reply:
Correct. I will fix this.